relays/ethereum/src/message_lane_loop.rs

```diff
@@ -0,0 +1,591 @@
+	pub best_peer: PeerHeaderId,
+}
+
+/// State of source client in one-way message lane.
+pub type SourceClientState<P> = ClientState<SourceHeaderIdOf<P>, TargetHeaderIdOf<P>>;
+
+/// State of target client in one-way message lane.
+pub type TargetClientState<P> = ClientState<TargetHeaderIdOf<P>, SourceHeaderIdOf<P>>;
+
+/// Both clients state.
+#[derive(Debug, Default)]
+pub struct ClientsState<P: MessageLane> {
+	/// Source client state.
+	pub source: Option<SourceClientState<P>>,
+	/// Target client state.
+	pub target: Option<TargetClientState<P>>,
+}
+
+/// Run message lane service loop.
+pub fn run<P: MessageLane>(
+	mut source_client: impl SourceClient<P>,
+	source_tick: Duration,
+	mut target_client: impl TargetClient<P>,
+	target_tick: Duration,
+	reconnect_delay: Duration,
+	stall_timeout: Duration,
+	exit_signal: impl Future<Output = ()>,
+) {
+	let mut local_pool = futures::executor::LocalPool::new();
+	let exit_signal = exit_signal.shared();
+
+	local_pool.run_until(async move {
+		loop {
+			let result = run_until_connection_lost(
+				source_client.clone(),
+				source_tick,
+				target_client.clone(),
+				target_tick,
+				stall_timeout,
+				exit_signal.clone(),
+			)
+			.await;
+
+			match result {
+				Ok(()) => break,
+				Err(failed_client) => {
+					async_std::task::sleep(reconnect_delay).await;
+					if failed_client == FailedClient::Both || failed_client == FailedClient::Source {
+						source_client = source_client.reconnect();
+					}
+					if failed_client == FailedClient::Both || failed_client == FailedClient::Target {
+						target_client = target_client.reconnect();
+					}
+				}
+			}
+
+			log::debug!(
+				target: "bridge",
+				"Restarting lane {} -> {}",
+				P::SOURCE_NAME,
+				P::TARGET_NAME,
+			);
+		}
+	});
+}
+
+/// Run one-way message delivery loop until connection with target or source node is lost, or exit signal is received.
+async fn run_until_connection_lost<P: MessageLane, SC: SourceClient<P>, TC: TargetClient<P>>(
+	source_client: SC,
+	source_tick: Duration,
+	target_client: TC,
+	target_tick: Duration,
+	stall_timeout: Duration,
+	exit_signal: impl Future<Output = ()>,
+) -> Result<(), FailedClient> {
+	let mut source_retry_backoff = retry_backoff();
+	let mut source_client_is_online = false;
+	let mut source_state_required = true;
+	let source_state = source_client.state().fuse();
+	let source_go_offline_future = futures::future::Fuse::terminated();
+	let source_tick_stream = interval(source_tick).fuse();
+
+	let mut target_retry_backoff = retry_backoff();
+	let mut target_client_is_online = false;
+	let mut target_state_required = true;
+	let target_state = target_client.state().fuse();
+	let target_go_offline_future = futures::future::Fuse::terminated();
+	let target_tick_stream = interval(target_tick).fuse();
+
+	let (
+		(delivery_source_state_sender, delivery_source_state_receiver),
+		(delivery_target_state_sender, delivery_target_state_receiver),
+	) = (unbounded(), unbounded());
```

Comment:
Do we not risk memory leaks with `unbounded` channels?

Reply:
How leaks may happen here? You probably mean high memory consumption? The only hypothetical case is when we fail to wake up on non-empty channel. Which should not happen because ready-futures-to-awake are selected in random order.